amazon/ion/reader_binary.py

```diff
@@ -208,406 +208,275 @@ def _parse_sid_iter(data):
+    parser: Callable[[_ParserContext], Tuple[IonEvent, SliceableBuffer]]
+    ion_type: Optional[IonType]
+    depth: int
+    limit: int
+
+
+@coroutine
+def stream_handler():
+    """
+    Handler for an Ion Binary value-stream.
+    """
+    buffer = SliceableBuffer.empty()
+    # top-level context limit is -1 to denote no limit
+    context_stack = deque([_ContextFrame(_tlv_parser, None, 0, -1)])
+    cursor = 0
+    ion_event = None
+    skip_or_next = ReadEventType.NEXT
+    expect_data = False
+    # will get swapped out for tlv or struct parser in main loop
+    parser = _ivm_parser
+    parent_type = None
+    depth = 0
+    limit = -1
+
+    # This is the main event loop for the parser coroutine.
+    #
+    # Each iteration begins with responding with previous ion_event (None to start)
+    # and receiving the user's read event.
+    #
+    # Then there are two main parts:
+    # 1/ handles the users' request, checking invariants and extending the
+    #    buffer and/or skipping.
+    # 2/ parses if possible then mutates state based on the results of parsing.
+    #
+    # You should think of them as distinct functions, inlined to avoid stack
+    # push/pop overhead.
     while True:
-        data_event, self = (yield transition)
-        if data_event is not None and data_event.type is ReadEventType.SKIP:
-            yield ctx.read_data_transition(ctx.remaining, self, skip=True)
-
-        if ctx.queue.position == ctx.limit:
-            # We are at the end of the container.
-            # Yield the close event and go to enclosing container.
-            yield Transition(
-                IonEvent(IonEventType.CONTAINER_END, ion_type, depth=ctx.depth-1),
-                ctx.whence
-            )
-
-        if ion_type is IonType.STRUCT:
-            # Read the field name.
-            self_handler = _create_delegate_handler(self)
-            (field_sid, _), _ = yield ctx.immediate_transition(
-                _var_uint_field_handler(self_handler, ctx)
-            )
-            field_name = SymbolToken(None, field_sid)
+        read_event = yield ion_event
+        assert read_event is not None
+
+        # part 1: handle user's read event
+        if expect_data:
+            if read_event.type is not ReadEventType.DATA:
+                raise TypeError("Data expected!")
+            buffer = buffer.extend(read_event.data)
         else:
-            field_name = None
+            if read_event.type is ReadEventType.DATA:
+                raise TypeError("Next or Skip expected!")
+            skip_or_next = read_event.type
+
+        ion_event = None
+        if skip_or_next is ReadEventType.SKIP:
+            if parent_type is None:
+                raise TypeError("Cannot Skip Top-level values!")
+
+            to_skip = limit - cursor
+            (skipped, buffer) = buffer.skip(to_skip)
+            cursor += skipped
+            if cursor < limit:
+                ion_event = ION_STREAM_INCOMPLETE_EVENT
```

Comment:
`IncompleteReadError` is used to convey this in the case of `read`. Should we be consistent?

Reply:
I'm assuming that you get that the buffer talks in terms of the exceptions and the parser loop talks in terms of Ion Events and this comment is saying the same as above just at call-site. LMK if I'm misinterpreting.

Reply:
Yes, it's the same comment as the above.